### DISCUSSION TITLE
Packet drop with Ikev1 host-to-host transport ipsec

### DISCUSSION THREAD

**Discussion by paolodoz** [strongswan/strongswan#1971] [Q&A] [answered]
Hello everybody,
I'm trying to configure an host to host ipsec transport connection (Ikev1 PSK phase 1 aes128-md5-modp1536 and phase 2 aes256-sha1 )
I'm forced to use these parameters to have compatibility with legacy company infrastructure.

Connection is correctly established, but when I try to exchange some data (i.e ping) I see the traffic going out, arriving to the other endpoint and then packet is dropped but I don't understand why.

Here after some configuration file and output of ip and other commands

Config file of embedded device

```
connections {
        ciscoios {
                version = 1
                local_addrs = 192.168.1.2
                remote_addrs = 192.168.1.34
                proposals = aes128-md5-modp1536
                aggressive = no
                local-up {
                        auth = psk
                        id = 192.168.1.2
                }
                remote-up {
                        auth = psk
                        id = 192.168.1.34
                }
                children {
                        net {
                                esp_proposals = aes256-sha1
                                mode = transport
                        }
                }
        }
}

secrets {
        ike-microup {
                id = 192.168.1.34
                secret = 0sFpZAZqEN6Ti9sqt4ZP5EWcqx
        }
}

``` 

host config
```
connections {
   embedded {
      local_addrs  = 192.168.1.34
      remote_addrs = 192.168.1.2
      local {
         auth = psk
      }
      remote {
         auth = psk
         id = 192.168.1.2
      }
      children {
         net {
            esp_proposals = aes256-sha1
	    mode = transport	
         }
      }
      version = 1
      proposals = aes128-md5-modp1536
   }
}

secrets {
   ike-embedded {
      id = 192.168.1.2
      secret = 0sFpZAZqEN6Ti9sqt4ZP5EWcqx
   }
}
``` 

Establishing connection from the device

> /etc/swanctl/conf.d # swanctl -i -c net
[IKE] initiating Main Mode IKE_SA ciscoios[2] to 192.168.1.34
[ENC] generating ID_PROT request 0 [ SA V V V V V ]
[NET] sending packet: from 192.168.1.2[500] to 192.168.1.34[500] (180 bytes)
[NET] received packet: from 192.168.1.34[500] to 192.168.1.2[500] (160 bytes)
[ENC] parsed ID_PROT response 0 [ SA V V V V ]
[IKE] received XAuth vendor ID
[IKE] received DPD vendor ID
[IKE] received FRAGMENTATION vendor ID
[IKE] received NAT-T (RFC 3947) vendor ID
[CFG] selected proposal: IKE:AES_CBC_128/HMAC_MD5_96/PRF_HMAC_MD5/MODP_1536
[ENC] generating ID_PROT request 0 [ KE No NAT-D NAT-D ]
[NET] sending packet: from 192.168.1.2[500] to 192.168.1.34[500] (300 bytes)
[NET] received packet: from 192.168.1.34[500] to 192.168.1.2[500] (300 bytes)
[ENC] parsed ID_PROT response 0 [ KE No NAT-D NAT-D ]
[ENC] generating ID_PROT request 0 [ ID HASH N(INITIAL_CONTACT) ]
[NET] sending packet: from 192.168.1.2[500] to 192.168.1.34[500] (92 bytes)
[NET] received packet: from 192.168.1.34[500] to 192.168.1.2[500] (76 bytes)
[ENC] parsed ID_PROT response 0 [ ID HASH ]
[IKE] IKE_SA ciscoios[2] established between 192.168.1.2[192.168.1.2]...192.168.1.34[192.168.1.34]
[IKE] scheduling rekeying in 14317s
[IKE] maximum IKE_SA lifetime 15757s
[ENC] generating QUICK_MODE request 1709850084 [ HASH SA No ID ID ]
[NET] sending packet: from 192.168.1.2[500] to 192.168.1.34[500] (172 bytes)
[NET] received packet: from 192.168.1.34[500] to 192.168.1.2[500] (172 bytes)
[ENC] parsed QUICK_MODE response 1709850084 [ HASH SA No ID ID ]
[CFG] selected proposal: ESP:AES_CBC_256/HMAC_SHA1_96/NO_EXT_SEQ
[IKE] CHILD_SA net{2} established with SPIs cdce349d_i cb5838e7_o and TS 192.168.1.2/32 === 192.168.1.34/32
initiate completed 
successfully

ping is not successful in both direction, but I see this
```
cat /proc/net/xfrm_stat 
....
XfrmInStateProtoError           13
....

``` 
```
/etc/swanctl/conf.d # ip -s xfrm state
src 192.168.1.2 dst 192.168.1.34
        proto esp spi 0xcb5838e7(3411556583) reqid 1(0x00000001) mode transport
        replay-window 0 seq 0x00000000 flag  (0x00000000)
        auth-trunc hmac(sha1) 0x00ee5d851f2c453d2fcbec283530be186ae5ecb8 (160 bits) 96
        enc cbc(aes) 0x4e7095e9686641baec7198ac67de776bfc7970465efcde989bc267a99c838667 (256 bits)
        anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
        sel src 192.168.1.2/32 dst 192.168.1.34/32 uid 0
        lifetime config:
          limit: soft (INF)(bytes), hard (INF)(bytes)
          limit: soft (INF)(packets), hard (INF)(packets)
          expire add: soft 3414(sec), hard 3960(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2023-10-31 13:32:06 use -
        stats:
          replay-window 0 replay 0 failed 0
src 192.168.1.34 dst 192.168.1.2
        proto esp spi 0xcdce349d(3452843165) reqid 1(0x00000001) mode transport
        replay-window 32 seq 0x00000000 flag  (0x00000000)
        auth-trunc hmac(sha1) 0xcd41568c5e39cc5dfee93fd7d596964d81055413 (160 bits) 96
        enc cbc(aes) 0x53a0ddf9ea53038689ad146b6a1bd6c0c4e843bb8e61920ee51f6b474f4a2d44 (256 bits)
        anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
        sel src 192.168.1.34/32 dst 192.168.1.2/32 uid 0
        lifetime config:
          limit: soft (INF)(bytes), hard (INF)(bytes)
          limit: soft (INF)(packets), hard (INF)(packets)
          expire add: soft 3347(sec), hard 3960(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2023-10-31 13:32:06 use -
        stats:
          replay-window 0 replay 0 failed 0
``` 
```
/etc/swanctl/conf.d # ip -s xfrm policy
src 192.168.1.2/32 dst 192.168.1.34/32 uid 0
        dir out action allow index 161 priority 367231 share any flag  (0x00000000)
        lifetime config:
          limit: soft (INF)(bytes), hard (INF)(bytes)
          limit: soft (INF)(packets), hard (INF)(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2023-10-31 13:32:06 use -
        tmpl src 0.0.0.0 dst 0.0.0.0
                proto esp spi 0xcb5838e7(3411556583) reqid 1(0x00000001) mode transport
                level required share any 
                enc-mask ffffffff auth-mask ffffffff comp-mask ffffffff
src 192.168.1.34/32 dst 192.168.1.2/32 uid 0
        dir in action allow index 152 priority 367231 share any flag  (0x00000000)
        lifetime config:
          limit: soft (INF)(bytes), hard (INF)(bytes)
          limit: soft (INF)(packets), hard (INF)(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2023-10-31 13:32:06 use -
        tmpl src 0.0.0.0 dst 0.0.0.0
                proto esp spi 0x00000000(0) reqid 1(0x00000001) mode transport
                level required share any 
                enc-mask ffffffff auth-mask ffffffff comp-mask ffffffff
src 0.0.0.0/0 dst 0.0.0.0/0 uid 0
        socket in action allow index 147 priority 0 share any flag  (0x00000000)
        lifetime config:
          limit: soft 0(bytes), hard 0(bytes)
          limit: soft 0(packets), hard 0(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2023-10-31 13:31:43 use -
src 0.0.0.0/0 dst 0.0.0.0/0 uid 0
        socket out action allow index 140 priority 0 share any flag  (0x00000000)
        lifetime config:
          limit: soft 0(bytes), hard 0(bytes)
          limit: soft 0(packets), hard 0(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2023-10-31 13:31:43 use -
src 0.0.0.0/0 dst 0.0.0.0/0 uid 0
        socket in action allow index 131 priority 0 share any flag  (0x00000000)
        lifetime config:
          limit: soft 0(bytes), hard 0(bytes)
          limit: soft 0(packets), hard 0(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2023-10-31 13:31:43 use 2023-10-31 13:32:06
src 0.0.0.0/0 dst 0.0.0.0/0 uid 0
        socket out action allow index 124 priority 0 share any flag  (0x00000000)
        lifetime config:
          limit: soft 0(bytes), hard 0(bytes)
          limit: soft 0(packets), hard 0(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2023-10-31 13:31:43 use 2023-10-31 13:32:06
src ::/0 dst ::/0 uid 0
        socket in action allow index 115 priority 0 share any flag  (0x00000000)
        lifetime config:
          limit: soft 0(bytes), hard 0(bytes)
          limit: soft 0(packets), hard 0(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2023-10-31 13:31:43 use -
src ::/0 dst ::/0 uid 0
        socket out action allow index 108 priority 0 share any flag  (0x00000000)
        lifetime config:
          limit: soft 0(bytes), hard 0(bytes)
          limit: soft 0(packets), hard 0(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2023-10-31 13:31:43 use -
src ::/0 dst ::/0 uid 0
        socket in action allow index 99 priority 0 share any flag  (0x00000000)
        lifetime config:
          limit: soft 0(bytes), hard 0(bytes)
          limit: soft 0(packets), hard 0(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2023-10-31 13:31:43 use -
src ::/0 dst ::/0 uid 0
        socket out action allow index 92 priority 0 share any flag  (0x00000000)
        lifetime config:
          limit: soft 0(bytes), hard 0(bytes)
          limit: soft 0(packets), hard 0(packets)
          expire add: soft 0(sec), hard 0(sec)
          expire use: soft 0(sec), hard 0(sec)
        lifetime current:
          0(bytes), 0(packets)
          add 2023-10-31 13:31:43 use -
```

Firewall disabled on both endpoints
```
/etc/swanctl/conf.d # iptables-save 
# Generated by iptables-save v1.8.7 on Tue Oct 31 14:42:36 2023
*filter
:INPUT ACCEPT [1822:149095]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [96:10465]
COMMIT
# Completed on Tue Oct 31 14:42:36 2023
```
As you can see, the interface is dropping packets
```
ethx1     Link encap:Ethernet  HWaddr 12:92:11:60:36:E6  
          inet addr:192.168.1.2  Bcast:192.168.1.255  Mask:255.255.255.0
          inet6 addr: fe80::bbd8:516b:d6b5:1f20/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:59593 errors:0 dropped:29201 overruns:0 frame:0
          TX packets:401 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:6177780 (5.8 MiB)  TX bytes:36703 (35.8 KiB)
```

With wireshark or tcpdump I see the incoming encrypted ESP packets but they are not decoded.

Do you have any suggestion on which could be the root cause of this behavior? After a long search I didn't find anything helpful. Also in tunnel mode I have the same behavior

Thanks
Paolo

**Comment by tobiasbrunner** [comment]
> I'm forced to use these parameters to have compatibility with legacy company infrastructure.

I don't understand. You clearly configured strongSwan on both ends! Never use IKEv1 for that. (Not that I'd expect it to make a difference in this particular case.)

> /etc/swanctl/conf.d # ip -s xfrm state

Neither SA is used there. Did you only send traffic from the other end? If so, was that the case here too? (i.e. the `failed` counter on the inbound SA never increases?)

> As you can see, the interface is dropping packets

This might be an issue but that counter is not related to IPsec processing.

> Do you have any suggestion on which could be the root cause of this behavior?

Not really. Might just be some issue in the kernel you are using.

**Comment by paolodoz** [comment]
Hello Tobias,
> > I'm forced to use these parameters to have compatibility with legacy company infrastructure.
> 
> I don't understand. You clearly configured strongSwan on both ends! Never use IKEv1 for that. (Not that I'd expect it to make a difference in this particular case.)
> 
At the moment I'm trying to simulate the final condition that's why I set up the two endpoint with strongswan. When I'll be able to exchange data, one of this will be substituted by the real router. My configuration should work or not? Forget about possible security issue

> > /etc/swanctl/conf.d # ip -s xfrm state
> 
> Neither SA is used there. Did you only send traffic from the other end? If so, was that the case here too? (i.e. the `failed` counter on the inbound SA never increases?)

For the moment I'm trying to do a ping. The behavior is the same on both sides. Packet out is encrypted, received and discarded
> 
> > As you can see, the interface is dropping packets
> 
> This might be an issue but that counter is not related to IPsec processing.
> 
> > Do you have any suggestion on which could be the root cause of this behavior?
> 
> Not really. Might just be some issue in the kernel you are using.

I verified on strongswan documentation and all needed kernel CONFIG are enabled.
 
I spent a lot of time but no other ideas on what to check next, or which part debug further

Thanks

**Reply by tobiasbrunner** [reply]
> My configuration should work or not?

Sure, see [this example for an IKEv1 transport mode connection](https://www.strongswan.org/testing/testresults/ikev1/host2host-transport/index.html).

> For the moment I'm trying to do a ping. The behavior is the same on both sides. Packet out is encrypted, received and discarded

So check the counters of these SAs and the use times of the policies. The output above doesn't show any use in either direction.

> I spent a lot of time but no other ideas on what to check next, or which part debug further

Maybe your system isn't using iptables but nftables?

**Reply by paolodoz** [reply]
Hello Tobias,
I made further tests. I substituted my embedded device with a notebook and the connection is working, ping is successful.

```
root@x1c6:~# ip -s xfrm state
src 192.168.1.34 dst 192.168.1.23
	proto esp spi 0xc284702e(3263459374) reqid 1(0x00000001) mode transport
	replay-window 0 seq 0x00000000 flag  (0x00000000)
	auth-trunc hmac(sha1) 0xff14aacc4039b8535d0f20c9dfbf30f2cbc166c8 (160 bits) 96
	enc cbc(aes) 0xe75d967be51a484e182c91d1ed4fd5eec0d13d662a6ac7125d3d077662684d52 (256 bits)
	anti-replay context: seq 0x0, oseq 0x8, bitmap 0x00000000
	sel src 192.168.1.34/32 dst 192.168.1.23/32 uid 0
	lifetime config:
	  limit: soft (INF)(bytes), hard (INF)(bytes)
	  limit: soft (INF)(packets), hard (INF)(packets)
	  expire add: soft 3339(sec), hard 3960(sec)
	  expire use: soft 0(sec), hard 0(sec)
	lifetime current:
	  512(bytes), 8(packets)
	  add 2023-11-06 17:54:20 use 2023-11-06 17:54:38
	stats:
	  replay-window 0 replay 0 failed 0
src 192.168.1.23 dst 192.168.1.34
	proto esp spi 0xc8f2e0cb(3371360459) reqid 1(0x00000001) mode transport
	replay-window 32 seq 0x00000000 flag  (0x00000000)
	auth-trunc hmac(sha1) 0x04e74960f6dae219d837a9517165a26b578b14dc (160 bits) 96
	enc cbc(aes) 0x74d8522b3fd55d009316d9cc866152bc96604b54e9d45acc325ee331f495790c (256 bits)
	anti-replay context: seq 0x8, oseq 0x0, bitmap 0x000000ff
	sel src 192.168.1.23/32 dst 192.168.1.34/32 uid 0
	lifetime config:
	  limit: soft (INF)(bytes), hard (INF)(bytes)
	  limit: soft (INF)(packets), hard (INF)(packets)
	  expire add: soft 3531(sec), hard 3960(sec)
	  expire use: soft 0(sec), hard 0(sec)
	lifetime current:
	  512(bytes), 8(packets)
	  add 2023-11-06 17:54:20 use 2023-11-06 17:54:38
	stats:
	  replay-window 0 replay 0 failed 0
```
Then I compared the output when connection is established with the board, and I noticed that anti-replay bitmap is 0x00000000 instead 0x000000FF when the communication is working (the stats report also 9 failed messages) 

```
root@x1c6:~# ip -s x s
src 192.168.1.34 dst 192.168.1.2
	proto esp spi 0xc8c40ec2(3368292034) reqid 1(0x00000001) mode transport
	replay-window 0 seq 0x00000000 flag  (0x00000000)
	auth-trunc hmac(sha1) 0x1ce09edef783a0ed31f633539c6ee7524eb4150d (160 bits) 96
	enc cbc(aes) 0x43ed3e65301b0aaee6771247f510defab289aef762abe29935173f9a5486ec0c (256 bits)
	anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
	sel src 192.168.1.34/32 dst 192.168.1.2/32 uid 0
	lifetime config:
	  limit: soft (INF)(bytes), hard (INF)(bytes)
	  limit: soft (INF)(packets), hard (INF)(packets)
	  expire add: soft 3342(sec), hard 3960(sec)
	  expire use: soft 0(sec), hard 0(sec)
	lifetime current:
	  0(bytes), 0(packets)
	  add 2023-11-06 18:53:03 use -
	stats:
	  replay-window 0 replay 0 failed 0
src 192.168.1.2 dst 192.168.1.34
	proto esp spi 0xcb4fc7b5(3411003317) reqid 1(0x00000001) mode transport
	replay-window 32 seq 0x00000000 flag  (0x00000000)
	auth-trunc hmac(sha1) 0x17d7972a6d6467f26c6b7fc79e63ef427c7c3704 (160 bits) 96
	enc cbc(aes) 0xbda41dbd0cfa900358530d113d03e256c5d59edc55090838d714b5abcd3e0927 (256 bits)
	anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
	sel src 192.168.1.2/32 dst 192.168.1.34/32 uid 0
	lifetime config:
	  limit: soft (INF)(bytes), hard (INF)(bytes)
	  limit: soft (INF)(packets), hard (INF)(packets)
	  expire add: soft 3508(sec), hard 3960(sec)
	  expire use: soft 0(sec), hard 0(sec)
	lifetime current:
	  0(bytes), 0(packets)
	  add 2023-11-06 18:53:03 use 2023-11-06 18:53:10
	stats:
	  replay-window 0 replay 0 failed 9
```
The bitmap of anti-replay strategy could have an impact?
I checked and no nftables are used on both endpoints

Thanks

**Reply by tobiasbrunner** [reply]
> The bitmap of anti-replay strategy could have an impact?

No, that simply shows the current state of the bitmap used for anti-replay protection. On the other platform, packets are dropped before the anti-replay check so no updates to the anti-replay state/bitmap. The `failed` counter is increased if the integrity check for a packet fails, which seems to indicate corrupted packets. Maybe try capturing ESP packets on both ends and comparing them and maybe verifying the checksum via Wireshark.

**Reply by paolodoz** [reply]
With wireshark I correctly decrypted ping packets, but the ESP ICV seems to be incorrect. Is related to kernel ip stack or strongswan?

**Reply by tobiasbrunner** [reply]
Where did you capture the packets? On the receiver or the sender? Or is the effect the same on both ends? strongSwan definitely can't influence that as the traffic is handled by the kernel. But if you captured traffic on the receiver (or both and the problem only occurs on the receiver) then the packets are corrupted somehow on the way from the sender to the receiver.

**Reply by paolodoz** [reply]
Hello Tobias, I found that if I change from sha256 to sha348 everything works. Tomorrow I'll do some low level debug till I find the root cause

**Reply by paolodoz** [reply, marked as answer]
Hello the root cause is the hw crypto engine which outputs wrong values. Disabling it, everything works as expected.
Thanks for the support

**Reply by tobiasbrunner** [reply]
OK. What kind of crypto engine is that?

**Reply by paolodoz** [reply]
It's integrated in the STM3215XX platform

**Reply by tobiasbrunner** [reply]
Thanks. Could this be the issue described [here](https://community.st.com/t5/stm32-mcus-security/how-to-enable-the-stm32h750vbt6-hw-crypto/m-p/165198/highlight/true#M2300) (manufacturing issue in 2020 causing faulty crypto feature)?
